app/__init__.py

```diff
@@ -93,40 +95,35 @@ def auth():
     if "auth" not in request.args:
         return Response("", status=200)
 
-    from .auth.gitlab_auth import GitlabUserToken
-    from .auth.jupyterhub_auth import JupyterhubUserToken
-    from .auth.renku_auth import RenkuCoreAuthHeaders
-
     auths = {
-        "gitlab": GitlabUserToken,
-        "jupyterhub": JupyterhubUserToken,
-        "renku": RenkuCoreAuthHeaders,
+        "gitlab": gitlab_auth.GitlabUserToken,
+        "jupyterhub": jupyterhub_auth.JupyterhubUserToken,
+        "renku": renku_auth.RenkuCoreAuthHeaders,
+        "renku-cli": cli_auth.RenkuCoreCLIAuthHeaders,
     }
 
-    auth_arg = request.args.get("auth")
-    auth = auths[auth_arg]()
-    headers = dict(request.headers)
-
     # Keycloak public key is not defined so error
     if current_app.config["OIDC_PUBLIC_KEY"] is None:
         response = json.dumps("Ooops, something went wrong internally.")
         return Response(response, status=500)
 
+    auth_arg = request.args.get("auth")
+    headers = dict(request.headers)
+
     try:
+        auth = auths[auth_arg]()
+
         # validate incoming authentication
         # it can either be in session-cookie or Authorization header
-        new_tokens = web.get_valid_token(headers)
-        if new_tokens:
-            headers["Authorization"] = "Bearer {}".format(
-                new_tokens.get("access_token")
-            )
+        is_cli_request = auth_arg == "renku-cli"
+        new_token, audience = web.get_valid_token(headers, is_cli_request)
+        if new_token:
+            headers["Authorization"] = f"Bearer {new_token}"
+
         if "Authorization" in headers and "Referer" in headers:
             allowed = False
-            origins = jwt.decode(
-                headers["Authorization"][7:],
-                current_app.config["OIDC_PUBLIC_KEY"],
-                algorithms="RS256",
-                audience=current_app.config["OIDC_CLIENT_ID"],
+            origins = decode_keycloak_jwt(
+                token=headers["Authorization"][7:], audience=audience
```

Comment:
Hmm, I think there is a bit of a misunderstanding here (not added by this PR but already in the existing code) about meaning of the `audience`. The audience is not the client, but it is the intended recipient of a token:
```
[...] it is important for the authorization server to include the identity of the 
intended recipients (the audience), typically a single resource server (or a 
list of resource servers), in the token.
```
So renku (or maybe better RenkuLab?) is the audience for tokens that are issued for all clients. I'd suggest for the time being we use `current_app.config["OIDC_CLIENT_ID"]` (which is actually renku) as an audience for all clients, even though the fact that we use a `client_id` as an audience is very misleading. This will change at the latest once the `renku-ui` will have it's own client application registered with keylcoak and gets the access token from keycloak directly.

Reply:
Actually, using `renku` audience makes life a bit easier. I've switched to using `current_app.config["OIDC_CLIENT_ID"]`.